admin/command_runner.go

```diff
@@ -252,61 +239,29 @@ func (r *CommandRunner) runAdminServer(ctx context.Context) error {
 	return nil
 }
 
-func (r *CommandRunner) processLoop(ctx context.Context) {
-	defer func() {
-		r.logger.Info().Msg("process loop shutting down")
+func (r *CommandRunner) runCommand(ctx context.Context, command string, data map[string]interface{}) error {
+	r.logger.Info().Str("command", command).Msg("received new command")
 
-		// cleanup uncompleted requests from the command queue
-		for command := range r.commandQ {
-			close(command.responseChan)
+	if validator := r.getValidator(command); validator != nil {
+		if validationErr := validator(data); validationErr != nil {
+			return status.Error(codes.InvalidArgument, validationErr.Error())
 		}
+	}
 
-		r.workersFinished.Done()
-	}()
-
-	r.workersStarted.Done()
-
-	for {
-		select {
-		case command, ok := <-r.commandQ:
-			if !ok {
-				return
-			}
-
-			r.logger.Info().Str("command", command.command).Msg("received new command")
-
-			var err error
-
-			if validator := r.getValidator(command.command); validator != nil {
-				if validationErr := validator(command.data); validationErr != nil {
-					err = status.Error(codes.InvalidArgument, validationErr.Error())
-					goto sendResponse
-				}
-			}
-
-			if handler := r.getHandler(command.command); handler != nil {
-				// TODO: we can probably merge the command context with the worker context
-				// using something like: https://github.com/teivah/onecontext
-				if handleErr := handler(command.ctx, command.data); handleErr != nil {
-					if errors.Is(handleErr, context.Canceled) {
-						err = status.Error(codes.Canceled, "client canceled")
-					} else if errors.Is(handleErr, context.DeadlineExceeded) {
-						err = status.Error(codes.DeadlineExceeded, "request timed out")
-					} else {
-						s, _ := status.FromError(handleErr)
-						err = s.Err()
-					}
-				}
+	if handler := r.getHandler(command); handler != nil {
+		if handleErr := handler(ctx, data); handleErr != nil {
+			if errors.Is(handleErr, context.Canceled) {
+				return status.Error(codes.Canceled, "client canceled")
+			} else if errors.Is(handleErr, context.DeadlineExceeded) {
```

Comment:
we can actually do without the `if-elseif-else` here since we have a `return` in each of those clauses